### PR TITLE
[Py3] ipa-server-install

### PR DESCRIPTION
**Pull request by MartinBasti**
This PR allows to install ipa server using PY3 (CA, no DNS, no KRA, no CA-less, no external CA) and uninstall it (ipa-server-install --uninstall)

Functionality depends on #427

### CODE REVIEW

File: ipaserver/install/dsinstance.py [hunk 1]
```diff
@@ -951,21 +951,19 @@ def add_hbac(self):
...
-        dmpwdfile = ""
-        admpwdfile = ""
 
-        try:
-            (dmpwdfd, dmpwdfile) = tempfile.mkstemp(dir=paths.VAR_LIB_IPA)
-            os.write(dmpwdfd, self.dm_password)
-            os.close(dmpwdfd)
+        NTF = tempfile.NamedTemporaryFile
+        dir = paths.VAR_LIB_IPA
+        with NTF("w", dir=dir) as dmpwdfile, NTF("w", dir=dir) as admpwdfile:
```
Comment (tiran) [inline ipaserver/install/dsinstance.py]: urks, the code is getting even less readable.

File: ipaserver/plugins/baseldap.py [hunk 2]
```diff
@@ -1036,8 +1037,8 @@ def process_attr_options(self, entry_attrs, dn, keys, options):
                     except ValueError:
                         if isinstance(delval, bytes):
                             # This is a Binary value, base64 encode it
-                            delval = unicode(base64.b64encode(delval))
-                        raise errors.AttrValueNotFound(attr=attr, value=delval)
+                            delval = base64.b64encode(delval).decode('ascii')
+                            raise errors.AttrValueNotFound(attr=attr, value=delval)
```
Comment (tiran) [inline ipaserver/plugins/baseldap.py]: Same issue as in PR #427

File: ipaserver/install/dsinstance.py (hunk 1 above)
Comment (MartinBasti) [inline ipaserver/install/dsinstance.py]: I wanted to avoid extra indentation, I can use full form with backlashes

File: ipaserver/plugins/baseldap.py (hunk 2 above)
Comment (MartinBasti) [inline ipaserver/plugins/baseldap.py]: I will rebase it, half of this PR is PR #427 , thats why it is WIP

File: ipaserver/install/dsinstance.py (hunk 1 above)
Comment (MartinBasti) [inline ipaserver/install/dsinstance.py]: Can be this?
```
        dir_ipa = paths.VAR_LIB_IPA
        with tempfile.NamedTemporaryFile("w", dir=dir_ipa) as dmpwdfile, \
                tempfile.NamedTemporaryFile("w", dir=dir_ipa) as admpwdfile:
            dmpwdfile.write(self.dm_password)
            dmpwdfile.flush()
```

File: ipapython/ipautil.py [hunk 3]
```diff
@@ -859,7 +859,7 @@ def ipa_generate_password(entropy_bits=256, uppercase=1, lowercase=1, digits=1,
     # NSS function sftk_newPinCheck() in nss/lib/softoken/fipstokn.c.
     for charclass_name in ['digits', 'uppercase', 'lowercase', 'special']:
         charclass = pwd_charsets[charclass_name]
-        todo_characters = req_classes[charclass_name]
+        todo_characters = req_classes[charclass_name] or 0
```
Comment (HonzaCholasta) [inline ipapython/ipautil.py]: Why is this necessary? `req_classes` is constructed from integer arguments, no?

File: ipaserver/secrets/kem.py [hunk 4]
```diff
@@ -130,17 +130,16 @@ def set_key(self, usage, principal, key):
...
-                    ('cn', name),
-                    ('ipaKeyUsage', RFC5280_USAGE_MAP[usage]),
-                    ('memberPrincipal', principal),
+            mods = [('objectClass', [b'nsContainer',
+                                     b'ipaKeyPolicy',
+                                     b'ipaPublicKeyObject',
+                                     b'groupOfPrincipals']),
+                    ('cn', name.encode('utf-8')),
+                    ('ipaKeyUsage', RFC5280_USAGE_MAP[usage].encode('utf-8')),
+                    ('memberPrincipal', principal.encode('utf-8')),
```
Comment (HonzaCholasta) [inline ipaserver/secrets/kem.py]: How hard would it be to refactor this to use `ipaldap` (which handles encoding for you)?

File: ipaserver/secrets/kem.py [hunk 5]
```diff
@@ -181,7 +180,7 @@ def __init__(self, config=None, ipaconf=paths.IPA_DEFAULT_CONF):
         self.realm = conf.get('global', 'realm')
         self.ldap_uri = config.get('ldap_uri', None)
         if self.ldap_uri is None:
-            self.ldap_uri = conf.get('global', 'ldap_uri', None)
+            self.ldap_uri = conf.get('global', 'ldap_uri', raw=True)
```
Comment (HonzaCholasta) [inline ipaserver/secrets/kem.py]: How hard would it be to refactor this to use `ipalib.config`?

File: ipapython/ipautil.py (hunk 3 above)
Comment (MartinBasti) [inline ipapython/ipautil.py]: No, it can contain None, None == character class will not be used

File: ipaserver/secrets/kem.py (hunk 4 above)
Comment (MartinBasti) [inline ipaserver/secrets/kem.py]: I saw that custodia related modules are not using ipalib, so I'm not sure if this is on purpose or just NIH, so I don't know. I'm afraid I dont have enough time to refactor it.

File: ipaserver/secrets/kem.py (hunk 5 above)
Comment (MartinBasti) [inline ipaserver/secrets/kem.py]: Or why not to use API when this is in ldap context? We have to find first why custodia related modules is not using api, ipalib

File: ipaserver/install/dsinstance.py (hunk 1 above)
Comment (HonzaCholasta) [inline ipaserver/install/dsinstance.py]: How about
```python
        ...
        def tempf():
            return tempfile.NamedTemporaryFile('w', dir=paths.VAR_LIB_IPA)
        with tempf() as dmpwdfile, tempf() as admpwdfile:
            ...
```
?

File: ipaserver/secrets/kem.py (hunk 4 above)
Comment (HonzaCholasta) [inline ipaserver/secrets/kem.py]: OK.

**Comment by HonzaCholasta** [comment]
Fixed upstream
master:
https://fedorahosted.org/freeipa/changeset/88b192a37ead1538e6d840c2f686f8d21a948542
https://fedorahosted.org/freeipa/changeset/2674a217acc2864a5fed98d7ef1e7031eae4f866
https://fedorahosted.org/freeipa/changeset/c27a46177c710fb18bf5b02beab4bd82c191a4bc
https://fedorahosted.org/freeipa/changeset/8660b9e96801a764e808ca69c3c14a4a019d4eb8
https://fedorahosted.org/freeipa/changeset/d4aa75d10582443b38447985c3fce8e65fcd48a6
https://fedorahosted.org/freeipa/changeset/f31d73b79aaa1746f7d32576658fcd4136870115
https://fedorahosted.org/freeipa/changeset/7fd36e4d3651f327a8c3a2f13b92a2a304352dfd
https://fedorahosted.org/freeipa/changeset/47f912e16ba6de2f3579de610b0d902cf3e621a2
https://fedorahosted.org/freeipa/changeset/488d01ced715929d47f6766a63b7d6c597125562
https://fedorahosted.org/freeipa/changeset/69072cb80f8c4b7f6eff0c7cdfe6545fe59ea7b5
https://fedorahosted.org/freeipa/changeset/dd119f8aadb13e95e4db43053bc36c70977b001e
